File: agent/graph.js
const { StateGraph, END, START } = require('@langchain/langgraph');
const { AgentStateAnnotation } = require('./state');
const { fetchTicketsNode } = require('./nodes/fetchTickets');
const { assignTicketsNode } = require('./nodes/assignTickets');
const { planAgentNode } = require('./nodes/planAgent');
const { frontendAgentNode } = require('./nodes/frontendAgent');
const { backendAgentNode } = require('./nodes/backendAgent');
const { testAgentNode, PASS_THRESHOLD } = require('./nodes/testAgent');
const { fixAgentNode } = require('./nodes/fixAgent');
const { devopsAgentNode } = require('./nodes/devopsAgent');

const MAX_RETRIES = 3;

// ─────────────────────────────────────────────
// Conditional edge routing functions
// ─────────────────────────────────────────────

/**
 * After planAgent: route to frontend, backend, or frontend (then backend).
 */
function routeAfterPlan(state) {
  if (!state.currentTicket) return 'devops';

  const workType = state.plan?.workType ?? 'fullstack';
  if (workType === 'backend') return 'backend';
  return 'frontend'; // frontend or fullstack: start with frontend
}

/**
 * After frontendAgent: if fullstack, go to backend; else go to test.
 */
function routeAfterFrontend(state) {
  const workType = state.plan?.workType ?? 'fullstack';
  return workType === 'fullstack' ? 'backend' : 'test';
}

/**
 * After testAgent: pass → devops, fail → fix (if retries left) or devops.
 */
function routeAfterTest(state) {
  const { testScore, retryCount } = state;
  if (testScore >= PASS_THRESHOLD) return 'devops';
  if ((retryCount ?? 0) >= MAX_RETRIES) {
    console.log(`  ⚠️  Max retries (${MAX_RETRIES}) reached — deploying anyway.`);
    return 'devops';
  }
  return 'fix';
}

// ─────────────────────────────────────────────
// Build the LangGraph StateGraph
// ─────────────────────────────────────────────

function buildGraph() {
  const graph = new StateGraph(AgentStateAnnotation)
    // ── Nodes ──────────────────────────────────────────────
    .addNode('fetchTickets', fetchTicketsNode)
    .addNode('assignTickets', assignTicketsNode)
    .addNode('planAgent', planAgentNode)
    .addNode('frontendAgent', frontendAgentNode)
    .addNode('backendAgent', backendAgentNode)
    .addNode('testAgent', testAgentNode)
    .addNode('fixAgent', fixAgentNode)
    .addNode('devopsAgent', devopsAgentNode)

    // ── Linear edges ───────────────────────────────────────
    .addEdge(START, 'fetchTickets')
    .addEdge('fetchTickets', 'assignTickets')
    .addEdge('assignTickets', 'planAgent')

    // ── Conditional: plan → frontend | backend ─────────────
    .addConditionalEdges('planAgent', routeAfterPlan, {
      frontend: 'frontendAgent',
      backend: 'backendAgent',
      devops: 'devopsAgent',
    })

    // ── Conditional: frontend → backend | test ─────────────
    .addConditionalEdges('frontendAgent', routeAfterFrontend, {
      backend: 'backendAgent',
      test: 'testAgent',
    })

    .addEdge('backendAgent', 'testAgent')

    // ── Conditional: test → devops | fix ───────────────────
    .addConditionalEdges('testAgent', routeAfterTest, {
      devops: 'devopsAgent',
      fix: 'fixAgent',
    })

    .addEdge('fixAgent', 'testAgent')
    .addEdge('devopsAgent', END);

  return graph.compile();
}

const compiledGraph = buildGraph();

/**
 * Run the full agent pipeline once.
 * @returns {Promise<object>} Final pipeline state summary
 */
async function runAgentPipeline() {
  console.log('\n' + '═'.repeat(60));
  console.log('  🤖 SMS Agent Pipeline — Starting');
  console.log('═'.repeat(60));

  const finalState = await compiledGraph.invoke({});

  console.log('\n' + '═'.repeat(60));
  console.log('  🏁 Pipeline Complete');
  console.log(`  Ticket:   ${finalState.currentTicket?.key ?? 'none'}`);
  console.log(`  Score:    ${finalState.testScore}/100`);
  console.log(`  Deploy:   ${finalState.deployUrl ?? 'n/a'}`);
  console.log(`  Retries:  ${finalState.retryCount ?? 0}`);
  if (finalState.error) console.log(`  Error:    ${finalState.error}`);
  console.log('═'.repeat(60) + '\n');

  return {
    ticketKey: finalState.currentTicket?.key,
    testScore: finalState.testScore,
    deployUrl: finalState.deployUrl,
    retryCount: finalState.retryCount,
    filesGenerated: finalState.generatedFiles?.length ?? 0,
    error: finalState.error,
  };
}

module.exports = { runAgentPipeline, compiledGraph };

File: python-backend/graph.py
from langgraph.graph import StateGraph, START, END

from state import AgentState
from nodes.fetch_tickets import fetch_tickets_node
from nodes.assign_tickets import assign_tickets_node
from nodes.plan_agent import plan_agent_node
from nodes.frontend_agent import frontend_agent_node
from nodes.backend_agent import backend_agent_node
from nodes.test_agent import test_agent_node, PASS_THRESHOLD
from nodes.fix_agent import fix_agent_node
from nodes.devops_agent import devops_agent_node

MAX_RETRIES = 3


def route_after_plan(state: AgentState) -> str:
    if not state.get("current_ticket"):
        return "devops"
    work_type = (state.get("plan") or {}).get("workType", "fullstack")
    return "backend" if work_type == "backend" else "frontend"


def route_after_frontend(state: AgentState) -> str:
    work_type = (state.get("plan") or {}).get("workType", "fullstack")
    return "backend" if work_type == "fullstack" else "test"


def route_after_test(state: AgentState) -> str:
    score = state.get("test_score") or 0
    retries = state.get("retry_count") or 0
    if score >= PASS_THRESHOLD:
        return "devops"
    if retries >= MAX_RETRIES:
        print(f"  ⚠️  Max retries ({MAX_RETRIES}) reached — deploying anyway.")
        return "devops"
    return "fix"


def build_graph():
    graph = StateGraph(AgentState)

    graph.add_node("fetchTickets", fetch_tickets_node)
    graph.add_node("assignTickets", assign_tickets_node)
    graph.add_node("planAgent", plan_agent_node)
    graph.add_node("frontendAgent", frontend_agent_node)
    graph.add_node("backendAgent", backend_agent_node)
    graph.add_node("testAgent", test_agent_node)
    graph.add_node("fixAgent", fix_agent_node)
    graph.add_node("devopsAgent", devops_agent_node)

    graph.add_edge(START, "fetchTickets")
    graph.add_edge("fetchTickets", "assignTickets")
    graph.add_edge("assignTickets", "planAgent")
    graph.add_conditional_edges("planAgent", route_after_plan,
                                {"frontend": "frontendAgent", "backend": "backendAgent", "devops": "devopsAgent"})
    graph.add_conditional_edges("frontendAgent", route_after_frontend,
                                {"backend": "backendAgent", "test": "testAgent"})
    graph.add_edge("backendAgent", "testAgent")
    graph.add_conditional_edges("testAgent", route_after_test,
                                {"devops": "devopsAgent", "fix": "fixAgent"})
    graph.add_edge("fixAgent", "testAgent")
    graph.add_edge("devopsAgent", END)

    return graph.compile()


compiled_graph = build_graph()


async def run_agent_pipeline() -> dict:
    print("\n" + "═" * 60)
    print("  🤖 SMS Python Agent Pipeline — Starting")
    print("═" * 60)

    initial_state: AgentState = {
        "tickets": [], "assigned_tickets": {}, "current_ticket": None,
        "plan": None, "generated_files": [], "test_score": 0,
        "test_issues": [], "retry_count": 0, "deploy_url": None, "error": None,
    }

    final_state = await compiled_graph.ainvoke(initial_state)

    ticket_key = (final_state.get("current_ticket") or {}).get("key", "none")
    print("\n" + "═" * 60)
    print("  🏁 Pipeline Complete")
    print(f"  Ticket:   {ticket_key}")
    print(f"  Score:    {final_state.get('test_score')}/100")
    print(f"  Deploy:   {final_state.get('deploy_url') or 'n/a'}")
    print(f"  Retries:  {final_state.get('retry_count') or 0}")
    if final_state.get("error"):
        print(f"  Error:    {final_state['error']}")
    print("═" * 60 + "\n")

    return {
        "ticket_key": ticket_key,
        "test_score": final_state.get("test_score"),
        "deploy_url": final_state.get("deploy_url"),
        "retry_count": final_state.get("retry_count") or 0,
        "files_generated": len(final_state.get("generated_files") or []),
        "error": final_state.get("error"),
    }
